controllers/drplacementcontrol_controller.go

```diff
@@ -0,0 +1,1719 @@
+			preferredCluster = d.instance.Status.PreferredDecision.ClusterName
+			preferredClusterNamespace = d.instance.Status.PreferredDecision.ClusterNamespace
+		}
+	}
+
+	return preferredCluster, preferredClusterNamespace
+}
+
+func (d *DRPCInstance) processVRGForSecondaries() (bool, error) {
+	const ensured = true
+
+	if len(d.userPlacementRule.Status.Decisions) != 0 {
+		// clear current user PlacementRule's decision
+		err := d.updateUserPlacementRule("", "")
+		if err != nil {
+			return !ensured, err
+		}
+	}
+
+	failedCount := 0
+
+	for _, cluster := range d.drpcPlacementRule.Spec.Clusters {
+		err := d.updateVRGStateToSecondary(cluster.Name)
+		if err != nil {
+			d.log.Error(err, "Failed to update VRG to secondary", "cluster", cluster.Name)
+
+			failedCount++
+		}
+	}
+
+	if failedCount != 0 {
+		d.log.Info("Failed to update VRG to secondary", "FailedCount", failedCount)
+
+		return !ensured, nil
+	}
+
+	for _, cluster := range d.drpcPlacementRule.Spec.Clusters {
+		if !d.ensureVRGIsSecondary(cluster.Name) {
+			d.log.Info("Still waiting for VRG to transition to secondary", "cluster", cluster.Name)
+
+			return !ensured, nil
+		}
+	}
+
+	return ensured, nil
+}
+
+// outputs a string for use in creating a ManagedClusterView name
+// example: when looking for a vrg with name 'demo' in the namespace 'ramen', input: ("demo", "ramen", "vrg")
+// this will give output "demo-ramen-vrg-mcv"
+func BuildManagedClusterViewName(resourceName, resourceNamespace, resource string) string {
+	return fmt.Sprintf("%s-%s-%s-mcv", resourceName, resourceNamespace, resource)
+}
+
+func (r *DRPlacementControlReconciler) getVRGFromManagedCluster(
+	resourceName string, resourceNamespace string, managedCluster string) (*rmn.VolumeReplicationGroup, error) {
+	// get VRG and verify status through ManagedClusterView
+	mcvMeta := metav1.ObjectMeta{
+		Name:      BuildManagedClusterViewName(resourceName, resourceNamespace, "vrg"),
+		Namespace: managedCluster,
+	}
+
+	mcvViewscope := fndv2.ViewScope{
+		Resource:  "VolumeReplicationGroup",
+		Name:      resourceName,
+		Namespace: resourceNamespace,
+	}
+
+	vrg := &rmn.VolumeReplicationGroup{}
+
+	err := r.getManagedClusterResource(mcvMeta, mcvViewscope, vrg)
+
+	return vrg, err
+}
+
+func (d *DRPCInstance) updateUserPlRuleAndCreateVRGMW(homeCluster, homeClusterNamespace string) error {
+	err := d.updateUserPlacementRule(homeCluster, homeClusterNamespace)
+	if err != nil {
+		return err
+	}
+
+	// Creating VRG ManifestWork will be running in parallel with ACM deploying app resources
+	err = d.createVRGManifestWork(homeCluster)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *DRPCInstance) createPVManifestWorkForRestore(newPrimary string) error {
+	pvMWName := BuildManifestWorkName(d.instance.Name, d.instance.Namespace, MWTypePV)
+
+	existAndApplied, err := d.isManifestExistAndApplied(pvMWName, newPrimary)
+	if err != nil && !errors.IsNotFound(err) {
+		d.log.Error(err, "actuall here")
```

Comment:
Can the error log be bit more informative? Helps in understanding the context while debugging.

Reply:
Good catch. That's actually was there for debugging.